### PR TITLE
Part 1 of 2 - Making concrete math transformers 

### PR DESCRIPTION
**Pull request by leahmcguire**
**Related issues**
Want to use concrete math transformer classes so they are easier to generate outside of the shortcuts

**Describe the proposed solution**
Move from lamda definitions of transformers to defined classes

**Additional context**
Next PR will add more math transformers

### CODE REVIEW

File: core/src/main/scala/com/salesforce/op/stages/impl/feature/MathTransformers.scala [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+ * @tparam I      input feature type
+ * @tparam N      value type
+ */
+class ScalarAddTransformer[I <: OPNumeric[_], N]
+(
+  val scalar: N,
+  uid: String = UID[ScalarAddTransformer[_, _]]
+)(
+  implicit override val tti: TypeTag[I],
+  n: Numeric[N]
```
Comment (tovbinm) [inline core/src/main/scala/com/salesforce/op/stages/impl/feature/MathTransformers.scala]: `val n: Numeric[N]` (here and in other transformers)

File: core/src/main/scala/com/salesforce/op/stages/impl/feature/MathTransformers.scala [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+ * Real(x)    + Real.empty = Real(x)
+ * Real(x)    + Real(y)    = Real(x + y)
+ */
+class AddTransformer[I1 <: OPNumeric[_], I2 <: OPNumeric[_]]
+(
+  uid: String = UID[AddTransformer[_, _]]
+)(
+  implicit override val tti1: TypeTag[I1],
+  override val tti2: TypeTag[I2]
+) extends BinaryTransformer[I1, I2, Real](operationName = "addition", uid = uid){
```
Comment (tovbinm) [inline core/src/main/scala/com/salesforce/op/stages/impl/feature/MathTransformers.scala]: the operation name was `"plus"` is there a reason to change it (same with other transformers). I would rather keep it the same as before.

File: core/src/test/scala/com/salesforce/op/dsl/RichNumericFeatureTest.scala [hunk 3]
```diff
@@ -134,15 +134,55 @@ class RichNumericFeatureTest extends FlatSpec with FeatureTestBase with RichNume
     // TODO: add vectorize() test
   }
 
-  Spec[RichRealNNFeature] should "have tests" in {
-    // TODO: add tests
+  Spec[RichRealNNFeature] should "perform math functions correctly" in {
```
Comment (tovbinm) [inline core/src/test/scala/com/salesforce/op/dsl/RichNumericFeatureTest.scala]: thanks for the tests!

File: core/src/main/scala/com/salesforce/op/stages/impl/feature/MathTransformers.scala [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+class ScalarDivideTransformer[I <: OPNumeric[_], N]
+(
+  val scalar: N,
+  uid: String = UID[ScalarDivideTransformer[_, _]]
+)(
+  implicit override val tti: TypeTag[I],
+  n: Numeric[N]
+) extends UnaryTransformer[I, Real](operationName = "scalarDivide", uid = uid){
+  override def transformFn: I => Real = (i: I) => i.toDouble.map(_ / n.toDouble(scalar)).filter(Number.isValid).toReal
+}
```
Comment (tovbinm) [inline core/src/main/scala/com/salesforce/op/stages/impl/feature/MathTransformers.scala]: you might need an endline here

**Comment by tovbinm** [comment]
@leahmcguire can you also please add a test for each of the transformers using the transformer test spec?

File: core/src/test/scala/com/salesforce/op/dsl/RichNumericFeatureTest.scala (hunk 3 above)
Comment (Jauntbox) [reply core/src/test/scala/com/salesforce/op/dsl/RichNumericFeatureTest.scala]: Is there an easy way to use our `OpTransformerSpec` here, or would we have to make a separate test class for each transformer we're trying to test?

**Pull request merged**